src/coreclr/jit/morph.cpp

```diff
@@ -11553,45 +11558,62 @@ GenTree* Compiler::fgMorphHWIntrinsic(GenTreeHWIntrinsic* tree)
                 {
                     // Move constant vectors from op1 to op2 for comparison operations
 
-                    var_types simdBaseType = tree->GetSimdBaseType();
-                    unsigned  simdSize     = tree->GetSimdSize();
+                    genTreeOps newOper = GenTree::SwapRelop(oper);
+                    var_types  lookupType =
+                        GenTreeHWIntrinsic::GetLookupTypeForCmpOp(this, newOper, retType, simdBaseType, simdSize);
+                    NamedIntrinsic newId =
+                        GenTreeHWIntrinsic::GetHWIntrinsicIdForCmpOp(this, newOper, retType, op2, op1, simdBaseType,
+                                                                     simdSize, false);
 
-                    genTreeOps     newOper = GenTree::SwapRelop(oper);
-                    NamedIntrinsic newId   = GenTreeHWIntrinsic::GetHWIntrinsicIdForCmpOp(this, newOper, op2, op1,
-                                                                                          simdBaseType, simdSize, false);
+                    if (newId != NI_Illegal)
+                    {
+                        tree->ResetHWIntrinsicId(newId, op2, op1);
 
-                    tree->ResetHWIntrinsicId(newId, op2, op1);
+                        if (lookupType != retType)
+                        {
+                            assert(varTypeIsMask(lookupType));
+                            tree->gtType = lookupType;
```

Comment:
Why is it ok to retype `tree` to a mask type here if it didn't start out as a mask type? Won't the parent node be expecting a non-mask type?

Reply:
Because we're inserting and returning the `CvtMaskToVector` below if the final tree produced by morph ends up mismatching.

We don't insert it here because we still want to try and fold and finish morphing the new comparison node in case other optimizations are available.

There's many morph paths that return new nodes, so this should be no different.

Reply:
-- It's inserted here: https://github.com/dotnet/runtime/pull/105883/files/92b5f01f4d490f84949f1afc3c9eb02678394b57#diff-5b83397bbbdd17bb9457998b520fdaaa474d165390985b66f32371561b6d0bacR11610-R11616

We make this transform, we try to fold, we try to optimize, and then we insert the `CvtMaskToVector` in the case the expected and actual return types mismatch.